File: signals/structure.py
from __future__ import annotations

import pandas as pd

from signals.common import atr
from signals.models import StructureSignal


def _confirmed_pivots(frame: pd.DataFrame, left: int = 3, right: int = 3) -> tuple[list[tuple[int, float]], list[tuple[int, float]]]:
    highs: list[tuple[int, float]] = []
    lows: list[tuple[int, float]] = []
    for idx in range(left, len(frame) - right):
        window = frame.iloc[idx - left : idx + right + 1]
        current_high = float(frame.iloc[idx]["high"])
        current_low = float(frame.iloc[idx]["low"])
        if current_high >= float(window["high"].max()):
            highs.append((idx, current_high))
        if current_low <= float(window["low"].min()):
            lows.append((idx, current_low))
    return highs, lows


def _latest_fvg(frame: pd.DataFrame) -> tuple[tuple[float, float] | None, bool]:
    latest_zone: tuple[float, float] | None = None
    mitigated = False
    for idx in range(2, len(frame)):
        c1 = frame.iloc[idx - 2]
        c3 = frame.iloc[idx]
        if float(c3["low"]) > float(c1["high"]):
            latest_zone = (float(c1["high"]), float(c3["low"]))
            current_low = float(frame.iloc[-1]["low"])
            current_high = float(frame.iloc[-1]["high"])
            mitigated = current_low <= latest_zone[1] and current_high >= latest_zone[0]
        elif float(c3["high"]) < float(c1["low"]):
            latest_zone = (float(c3["high"]), float(c1["low"]))
            current_low = float(frame.iloc[-1]["low"])
            current_high = float(frame.iloc[-1]["high"])
            mitigated = current_low <= latest_zone[1] and current_high >= latest_zone[0]
    return latest_zone, mitigated


def analyze_market_structure(
    frame: pd.DataFrame,
    *,
    ticker: str,
    market: str,
    timeframe: str,
    pivot_left: int = 3,
    pivot_right: int = 3,
    equal_tolerance_pct: float = 0.2,
) -> StructureSignal:
    working = frame.copy().sort_values("datetime").reset_index(drop=True)
    if working.empty or len(working) < pivot_left + pivot_right + 10:
        return StructureSignal(
            ticker=ticker,
            market=market,
            timeframe=timeframe,
            direction="neutral",
            structure_type="insufficient_data",
            key_levels={},
            explanation="Not enough candles to confirm market structure.",
        )

    highs, lows = _confirmed_pivots(working, pivot_left, pivot_right)
    latest_close = float(working.iloc[-1]["close"])
    atr_series = atr(working, 14)
    atr_value = float(atr_series.iloc[-1]) if pd.notna(atr_series.iloc[-1]) else 0.0
    previous_high = highs[-2][1] if len(highs) >= 2 else None
    previous_low = lows[-2][1] if len(lows) >= 2 else None
    current_swing_high = highs[-1][1] if highs else None
    current_swing_low = lows[-1][1] if lows else None

    bullish_bos = previous_high is not None and latest_close > previous_high
    bearish_bos = previous_low is not None and latest_close < previous_low
    choch = False
    if len(highs) >= 2 and len(lows) >= 2:
        choch = (highs[-1][1] < highs[-2][1] and lows[-1][1] < lows[-2][1] and bullish_bos) or (
            highs[-1][1] > highs[-2][1] and lows[-1][1] > lows[-2][1] and bearish_bos
        )
    equal_highs = bool(len(highs) >= 2 and previous_high and abs(highs[-1][1] - previous_high) / previous_high * 100 <= equal_tolerance_pct)
    equal_lows = bool(len(lows) >= 2 and previous_low and abs(lows[-1][1] - previous_low) / previous_low * 100 <= equal_tolerance_pct)
    liquidity_sweep = bool(
        previous_high is not None and float(working.iloc[-1]["high"]) > previous_high and latest_close < previous_high
    ) or bool(
        previous_low is not None and float(working.iloc[-1]["low"]) < previous_low and latest_close > previous_low
    )
    inducement = bool(len(lows) >= 3 and previous_low is not None and lows[-1][1] > lows[-2][1] > lows[-3][1])

    direction = "bullish" if bullish_bos else "bearish" if bearish_bos else "neutral"
    structure_type = "BOS" if bullish_bos or bearish_bos else "CHOCH" if choch else "range"

    order_block_zone = None
    if bullish_bos:
        prior_red = working.iloc[:-1].loc[working.iloc[:-1]["close"] < working.iloc[:-1]["open"]]
        if not prior_red.empty:
            candle = prior_red.iloc[-1]
            order_block_zone = (float(candle["low"]), float(candle["high"]))
    elif bearish_bos:
        prior_green = working.iloc[:-1].loc[working.iloc[:-1]["close"] > working.iloc[:-1]["open"]]
        if not prior_green.empty:
            candle = prior_green.iloc[-1]
            order_block_zone = (float(candle["low"]), float(candle["high"]))

    fvg_zone, fvg_mitigated = _latest_fvg(working)
    bos_level = previous_high if bullish_bos else previous_low if bearish_bos else current_swing_high or current_swing_low
    retest_zone = None
    if bos_level is not None and atr_value:
        retest_zone = (bos_level - atr_value * 0.25, bos_level + atr_value * 0.25)

    explanation_bits = [
        f"Latest confirmed structure is {structure_type} with {direction} bias.",
        "Liquidity sweep detected." if liquidity_sweep else "No fresh sweep on the latest bar.",
        "Fresh FVG available." if fvg_zone and not fvg_mitigated else "No fresh unmitigated FVG.",
    ]
    return StructureSignal(
        ticker=ticker,
        market=market,
        timeframe=timeframe,
        direction=direction,
        structure_type=structure_type,
        key_levels={
            "previous_high": previous_high,
            "previous_low": previous_low,
            "current_swing_high": current_swing_high,
            "current_swing_low": current_swing_low,
            "bos_level": bos_level,
        },
        retest_zone=retest_zone,
        invalidation_level=current_swing_low if direction == "bullish" else current_swing_high,
        bos=bullish_bos or bearish_bos,
        choch=choch,
        liquidity_sweep=liquidity_sweep,
        equal_highs=equal_highs,
        equal_lows=equal_lows,
        inducement=inducement,
        order_block_zone=order_block_zone,
        fvg_zone=fvg_zone,
        fvg_mitigated=fvg_mitigated,
        explanation=" ".join(explanation_bits),
    )

File: stocker/lib/engine/structure.ts
/**
 * Market-structure analysis (BOS / CHOCH, pivots, FVG, order block, liquidity
 * sweep). Ported from STOPICK signals/structure.py.
 */
import type { Bar, Direction, Market, StructureSignal, Timeframe } from "./types";
import { atr, last } from "./indicators";

function confirmedPivots(bars: Bar[], left = 3, right = 3): { highs: [number, number][]; lows: [number, number][] } {
  const highs: [number, number][] = [];
  const lows: [number, number][] = [];
  for (let idx = left; idx < bars.length - right; idx++) {
    const window = bars.slice(idx - left, idx + right + 1);
    const ch = bars[idx].high;
    const cl = bars[idx].low;
    if (ch >= Math.max(...window.map((b) => b.high))) highs.push([idx, ch]);
    if (cl <= Math.min(...window.map((b) => b.low))) lows.push([idx, cl]);
  }
  return { highs, lows };
}

function latestFvg(bars: Bar[]): { zone: [number, number] | null; mitigated: boolean } {
  let zone: [number, number] | null = null;
  let mitigated = false;
  const lastBar = bars[bars.length - 1];
  for (let idx = 2; idx < bars.length; idx++) {
    const c1 = bars[idx - 2];
    const c3 = bars[idx];
    if (c3.low > c1.high) {
      zone = [c1.high, c3.low];
      mitigated = lastBar.low <= zone[1] && lastBar.high >= zone[0];
    } else if (c3.high < c1.low) {
      zone = [c3.high, c1.low];
      mitigated = lastBar.low <= zone[1] && lastBar.high >= zone[0];
    }
  }
  return { zone, mitigated };
}

export function analyzeMarketStructure(
  bars: Bar[],
  opts: { ticker: string; market: Market; timeframe: Timeframe; pivotLeft?: number; pivotRight?: number; equalTolerancePct?: number },
): StructureSignal {
  const { ticker, market, timeframe } = opts;
  const pivotLeft = opts.pivotLeft ?? 3;
  const pivotRight = opts.pivotRight ?? 3;
  const equalTolerancePct = opts.equalTolerancePct ?? 0.2;

  const base: StructureSignal = {
    ticker,
    market,
    timeframe,
    direction: "neutral",
    structureType: "insufficient_data",
    keyLevels: {},
    retestZone: null,
    invalidationLevel: null,
    bos: false,
    choch: false,
    liquiditySweep: false,
    equalHighs: false,
    equalLows: false,
    inducement: false,
    orderBlockZone: null,
    fvgZone: null,
    fvgMitigated: false,
    explanation: "Not enough candles to confirm market structure.",
  };

  if (bars.length < pivotLeft + pivotRight + 10) return base;

  const { highs, lows } = confirmedPivots(bars, pivotLeft, pivotRight);
  const lastBar = bars[bars.length - 1];
  const latestClose = lastBar.close;
  const atrValue = last(atr(bars, 14), 0) || 0;

  const previousHigh = highs.length >= 2 ? highs[highs.length - 2][1] : null;
  const previousLow = lows.length >= 2 ? lows[lows.length - 2][1] : null;
  const currentSwingHigh = highs.length ? highs[highs.length - 1][1] : null;
  const currentSwingLow = lows.length ? lows[lows.length - 1][1] : null;

  const bullishBos = previousHigh !== null && latestClose > previousHigh;
  const bearishBos = previousLow !== null && latestClose < previousLow;

  let choch = false;
  if (highs.length >= 2 && lows.length >= 2) {
    const hLast = highs[highs.length - 1][1];
    const hPrev = highs[highs.length - 2][1];
    const lLast = lows[lows.length - 1][1];
    const lPrev = lows[lows.length - 2][1];
    choch = (hLast < hPrev && lLast < lPrev && bullishBos) || (hLast > hPrev && lLast > lPrev && bearishBos);
  }

  const equalHighs =
    highs.length >= 2 && previousHigh !== null && previousHigh !== 0
      ? (Math.abs(highs[highs.length - 1][1] - previousHigh) / previousHigh) * 100 <= equalTolerancePct
      : false;
  const equalLows =
    lows.length >= 2 && previousLow !== null && previousLow !== 0
      ? (Math.abs(lows[lows.length - 1][1] - previousLow) / previousLow) * 100 <= equalTolerancePct
      : false;

  const liquiditySweep =
    (previousHigh !== null && lastBar.high > previousHigh && latestClose < previousHigh) ||
    (previousLow !== null && lastBar.low < previousLow && latestClose > previousLow);

  const inducement =
    lows.length >= 3 &&
    previousLow !== null &&
    lows[lows.length - 1][1] > lows[lows.length - 2][1] &&
    lows[lows.length - 2][1] > lows[lows.length - 3][1];

  const direction: Direction = bullishBos ? "bullish" : bearishBos ? "bearish" : "neutral";
  const structureType = bullishBos || bearishBos ? "BOS" : choch ? "CHOCH" : "range";

  let orderBlockZone: [number, number] | null = null;
  if (bullishBos) {
    for (let i = bars.length - 2; i >= 0; i--) {
      if (bars[i].close < bars[i].open) {
        orderBlockZone = [bars[i].low, bars[i].high];
        break;
      }
    }
  } else if (bearishBos) {
    for (let i = bars.length - 2; i >= 0; i--) {
      if (bars[i].close > bars[i].open) {
        orderBlockZone = [bars[i].low, bars[i].high];
        break;
      }
    }
  }

  const { zone: fvgZone, mitigated: fvgMitigated } = latestFvg(bars);
  const bosLevel = bullishBos ? previousHigh : bearishBos ? previousLow : currentSwingHigh ?? currentSwingLow;
  let retestZone: [number, number] | null = null;
  if (bosLevel !== null && bosLevel !== undefined && atrValue) {
    retestZone = [bosLevel - atrValue * 0.25, bosLevel + atrValue * 0.25];
  }

  const explanation = [
    `Latest confirmed structure is ${structureType} with ${direction} bias.`,
    liquiditySweep ? "Liquidity sweep detected." : "No fresh sweep on the latest bar.",
    fvgZone && !fvgMitigated ? "Fresh FVG available." : "No fresh unmitigated FVG.",
  ].join(" ");

  return {
    ticker,
    market,
    timeframe,
    direction,
    structureType,
    keyLevels: {
      previous_high: previousHigh,
      previous_low: previousLow,
      current_swing_high: currentSwingHigh,
      current_swing_low: currentSwingLow,
      bos_level: bosLevel ?? null,
    },
    retestZone,
    invalidationLevel: direction === "bullish" ? currentSwingLow : currentSwingHigh,
    bos: bullishBos || bearishBos,
    choch,
    liquiditySweep,
    equalHighs,
    equalLows,
    inducement,
    orderBlockZone,
    fvgZone,
    fvgMitigated,
    explanation,
  };
}
